src/internal/storage/fileset/validator.go

```diff
@@ -0,0 +1,66 @@
+package fileset
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
+)
+
+var _ FileSet = &pathValidator{}
+
+type pathValidator struct {
+	fs FileSet
+}
+
+// NewPathValidator checks if there is collision between file and dir names.
+func NewPathValidator(fs FileSet) FileSet {
+	pf := &pathValidator{
+		fs: fs,
+	}
+	return pf
+}
+
+func (pf *pathValidator) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
+	ctx = pctx.Child(ctx, "pathIterator")
+	var prev *index.Index
+	var retErr error = nil
+	if err := pf.fs.Iterate(ctx, func(f File) error {
+		idx := f.Index()
+		if retErr == nil {
+			retErr = CheckIndex(prev, idx)
+		}
+		prev = idx
+		return cb(f)
+	}, opts...); err != nil {
+		return errors.EnsureStack(err)
+	}
+	if retErr != nil {
+		return errors.EnsureStack(retErr)
+	}
+	return nil
+}
+
+func (pf *pathValidator) IterateDeletes(_ context.Context, _ func(File) error, opts ...index.Option) error {
+	return errors.Errorf("iterating deletes in an path validator file set")
```

Comment:
nit, an -> a, same for shards below.

Reply:
Ah, I missed these when refactoring index filtering. I'll fix these!